### PR TITLE
Add base Continuous Integration pipeline

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

We need a way to automate Docker production image build, validation & publication.

## Proposal

Add base CircleCI workflow to:

- [x] build production & development Docker images
- [x] check the production image
- [x] publish the production image to DockerHub upon git tags or PR merges

### CODE REVIEW

**Comment by jmaupetit** [comment]
Friendly reminder: this PR is ready for review :stuck_out_tongue: @sampaccoud // @lunika

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+          requires:
+            - build
+          filters:
+            tags:
+              only: /.*/
+
+      # DockerHub publication.
+      #
+      # Publish docker images only if all build, lint and test jobs succeed and
+      # it has been tagged with a tag starting with the letter v
```
Comment (sampaccoud) [inline .circleci/config.yml]: ... or is on the master branch.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            sudo sysctl -w vm.max_map_count=262144
+            sudo sysctl vm.max_map_count
+      # Run back-end (Django) check
+      #
+      # Nota bene: to run the django app, we need to ensure that both PostgreSQL
+      # and ElasticSearch services are up and ready. To achieve this, we wrap
+      # commands execution with dockerize, a tiny tool installed in the CI
+      # image. In our case, dockerize will wait up to one minute that both the
+      # database and elastisearch containers opened their expected tcp port
+      # (5432 and 9200 resp.).
```
Comment (lunika) [inline .circleci/config.yml]: All this comment is I think wrong. You are not using `dockerise` with `BUILD_TARGET` set to `production`. And we don;t have a `CI` image.

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            docker images | grep -E "^fundocker/funmooc\s*(${DOCKER_TAG}.*|latest|master)"
+      # Publish images to DockerHub
+      #
+      # Nota bene: logged user (see "Login to DockerHub" step) must have write
+      # permission for the project's repository; this also implies that the
+      # DockerHub repository already exists.
+      - run:
+          name: Publish images
+          command: |
+            DOCKER_TAG=$([[ -z "$CIRCLE_TAG" ]] && echo $CIRCLE_BRANCH || echo ${CIRCLE_TAG} | sed 's/^v//')
```
Comment (lunika) [inline .circleci/config.yml]: Did you try to put this in the task's `environment` section ? If it's possible to compute the value in one place I think we should do it.

File: env.d/production [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Nota bene
```
Comment (lunika) [inline env.d/production]: It's weird to name this file `production` and only use it in the CI. I know it's to fit with what is done in the `docker-compose` file, but it's confusing IMO.

File: .circleci/config.yml (hunk 2 above)
Comment (sampaccoud) [reply .circleci/config.yml]: I nearly commented the same but in fact I now think he is doing what this comment says, only via the make file.

File: env.d/production (hunk 4 above)
Comment (jmaupetit) [reply env.d/production]: I know, this is why I wrote this preamble... I have no better option.

File: .circleci/config.yml (hunk 2 above)
Comment (jmaupetit) [reply .circleci/config.yml]: In fact, @lunika is right: this is why I have added a `sleep 10` command :sweat_smile:

File: .circleci/config.yml (hunk 3 above)
Comment (jmaupetit) [reply .circleci/config.yml]: I've defined environment variables. Let's see how it goes.

**Pull request merged**